### PR TITLE
Example of a custom ValidationAttribute

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
Add an example of custom `ValidationAttribute` localization to` Mvc.LocalizationSample.Web`.

### CODE REVIEW

File: samples/Mvc.LocalizationSample.Web/CustomValidationAttributeAdapterProvider.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+using Microsoft.AspNetCore.Mvc.DataAnnotations.Internal;
+using Microsoft.Extensions.Localization;
+
+namespace Mvc.LocalizationSample.Web
+{
+    public class CustomValidationAttributeAdapterProvider : IValidationAttributeAdapterProvider
+    {
+        public IAttributeAdapter GetAttributeAdapter(ValidationAttribute attribute, IStringLocalizer stringLocalizer)
+        {
+            var defaultProvider = new ValidationAttributeAdapterProvider();
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/CustomValidationAttributeAdapterProvider.cs]: This approach couples the replacement provider with MVC `.Internal` implementation details. Suggest instead registering a `Func` which calls `serviceProvider.GetRequiredService<IValidationAttributeAdapterProvider>()` and passes that to a new instance of this provider.

File: samples/Mvc.LocalizationSample.Web/NotEqualToAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            return string.Format(CultureInfo.CurrentCulture, ErrorMessageString, name, Convert.ToString(Value));
+        }
+
+        public override bool IsValid(object value)
+        {
+            return !string.Equals(Convert.ToString(Value), Convert.ToString(value), StringComparison.CurrentCultureIgnoreCase);
+        }
+
+        public object Value { get; set; }
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/NotEqualToAttribute.cs]: - No reason for this to be settable.
- Move above the methods.

File: samples/Mvc.LocalizationSample.Web/NotEqualToAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        public override string FormatErrorMessage(string name)
+        {
+            return string.Format(CultureInfo.CurrentCulture, ErrorMessageString, name, Convert.ToString(Value));
+        }
+
+        public override bool IsValid(object value)
+        {
+            return !string.Equals(Convert.ToString(Value), Convert.ToString(value), StringComparison.CurrentCultureIgnoreCase);
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/NotEqualToAttribute.cs]: Could simplify this and `FormatErrorMessage` doing something like [`[Phone]`](https://github.com/dotnet/corefx/blob/master/src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/PhoneAttribute.cs): If the value cannot be cast to `string`, it's not valid.

File: samples/Mvc.LocalizationSample.Web/NotEqualToAttributeAdapter.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Mvc.DataAnnotations.Internal;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
+using Microsoft.Extensions.Localization;
+
+namespace Mvc.LocalizationSample.Web
+{
+    internal class NotEqualToAttributeAdapter : AttributeAdapterBase<NotEqualToAttribute>
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/NotEqualToAttributeAdapter.cs]: Why would an app ever have an `internal` class?

File: samples/Mvc.LocalizationSample.Web/NotEqualToAttributeAdapter.cs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        }
+
+        public override void AddValidation(ClientModelValidationContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            MergeAttribute(context.Attributes, "data-val", "true");
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/NotEqualToAttributeAdapter.cs]: What client-side validation will this actually enable? I suspect some JavaScript code is missing from this sample for the scenario to work end-to-end.

Comment (ryanbrandenburg) [inline samples/Mvc.LocalizationSample.Web/NotEqualToAttributeAdapter.cs]: There is no JavaScript and therefore no client-side validation in the existing samples, this is just to make sure that anybody looking at this knows that that's the pattern. I've refactored it a bit to give a simpler idea of what you should do at the MergeAttribute part.

**Comment by ryanbrandenburg** [comment]
:up::date:

File: samples/Mvc.LocalizationSample.Web/CustomValidationAttributeAdapterProvider.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+        IAttributeAdapter IValidationAttributeAdapterProvider.GetAttributeAdapter(
+            ValidationAttribute attribute,
+            IStringLocalizer stringLocalizer)
+        {
+            var adapter = GetAttributeAdapter(attribute, stringLocalizer);
+
+            if (adapter == null)
+            {
+                var type = attribute.GetType();
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/CustomValidationAttributeAdapterProvider.cs]: Why is the `Type` needed? Suggest
``` c#
var minLengthSix = attribute as minLengthSix;
if (minLengthSix != null)
{
    adapter = new MinLengthSixAdapter(minLengthSix, stringLocalizer);
}
```

File: samples/Mvc.LocalizationSample.Web/CustomValidationAttributeAdapterProvider.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        public CustomValidationAttributeAdapterProvider()
+        {
+        }
+
+        IAttributeAdapter IValidationAttributeAdapterProvider.GetAttributeAdapter(
+            ValidationAttribute attribute,
+            IStringLocalizer stringLocalizer)
+        {
+            var adapter = GetAttributeAdapter(attribute, stringLocalizer);
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/CustomValidationAttributeAdapterProvider.cs]: Use `base.` to make it obvious why this doesn't recurse.

File: samples/Mvc.LocalizationSample.Web/CustomValidationAttributeAdapterProvider.cs [hunk 8]
```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.ComponentModel.DataAnnotations;
+using Microsoft.AspNetCore.Mvc.DataAnnotations;
+using Microsoft.AspNetCore.Mvc.DataAnnotations.Internal;
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/CustomValidationAttributeAdapterProvider.cs]: Please open an issue about moving `ValidationAttributeAdapterProvider` and `AttributeAdapterBase<TAttribute>` out of `.Internal`

File: samples/Mvc.LocalizationSample.Web/MinLengthSixAttribute.cs [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        }
+
+        public override bool IsValid(object value)
+        {
+            var stringValue = value as string;
+            if (stringValue == null)
+            {
+                return false;
+            }
+            else
```
Comment (dougbu) [inline samples/Mvc.LocalizationSample.Web/MinLengthSixAttribute.cs]: nit: No need for `else`

**Comment by ryanbrandenburg** [comment]
:up::date:

**Pull request merged**